source/linux/io_epoll_context.cpp

```diff
@@ -0,0 +1,485 @@
+                    : "acquired items from remote queue");
+  schedule_local(std::move(items));
+}
+
+bool io_epoll_context::try_register_remote_queue_notification() noexcept {
+  auto queuedItems = remoteQueue_.try_mark_inactive_or_dequeue_all();
+  LOG(queuedItems.empty() ? "remote queue is empty"
+                          : "registered items from remote queue");
+  if (!queuedItems.empty()) {
+    schedule_local(std::move(queuedItems));
+    return false;
+  }
+  return true;
+}
+
+void io_epoll_context::signal_remote_queue() {
+  LOG("writing bytes to eventfd");
+
+  // Notify eventfd() by writing a 64-bit integer to it.
+  const __u64 value = 1;
+  ssize_t bytesWritten =
+      write(remoteQueueEventFd_.get(), &value, sizeof(value));
+  if (bytesWritten < 0) {
+    // What to do here? Terminate/abort/ignore?
+    // Try to dequeue the item before returning?
+    LOG("error writing to remote queue eventfd");
+    int errorCode = errno;
+    throw std::system_error{errorCode, std::system_category()};
+  }
+
+  assert(bytesWritten == sizeof(value));
+}
+
+void io_epoll_context::remove_timer(schedule_at_operation* op) noexcept {
+  LOGX("remove_timer(%p)\n", (void*)op);
+
+  assert(!timers_.empty());
+  if (timers_.top() == op) {
+    timersAreDirty_ = true;
+  }
+  timers_.remove(op);
+}
+
+void io_epoll_context::update_timers() noexcept {
+  LOG("update_timers()");
+  // Reap any elapsed timers.
+  if (!timers_.empty()) {
+    time_point now = monotonic_clock::now();
+    while (!timers_.empty() && timers_.top()->dueTime_ <= now) {
+      schedule_at_operation* item = timers_.pop();
+
+      LOGX("dequeued elapsed timer %p\n", (void*)item);
+
+      if (item->canBeCancelled_) {
+        auto oldState = item->state_.fetch_add(
+            schedule_at_operation::timer_elapsed_flag,
+            std::memory_order_acq_rel);
+        if ((oldState & schedule_at_operation::cancel_pending_flag) != 0) {
+          LOGX("timer already cancelled\n");
+
+          // Timer has been cancelled by a remote thread.
+          // The other thread is responsible for enqueueing is operation onto
+          // the remoteQueue_.
+          continue;
+        }
+      }
+
+      // Otherwise, we are responsible for enqueuing the timer onto the
+      // ready-to-run queue.
+      schedule_local(item);
+    }
+  }
+
+  // Check if we need to cancel or start some new OS timers.
+  if (timers_.empty()) {
+    if (currentDueTime_.has_value()) {
+      LOG("no more schedule_at requests, cancelling timer");
+
+      // Cancel the outstanding timer.
+      if (try_submit_timer_io_cancel()) {
+        currentDueTime_.reset();
+        timersAreDirty_ = false;
+      }
+    }
+  } else {
+    time_point now = monotonic_clock::now();
+    const auto earliestDueTime = timers_.top()->dueTime_;
+    LOGX(
+      "next timer in %i ms\n",
+      (int)std::chrono::duration_cast<std::chrono::milliseconds>(
+          earliestDueTime - now)
```

Comment:
Consider inlining the call to `monotonic_clock::now()` here to avoid the extra call in the case that logging is disabled.
The `now` variable is not used outside of the logging statement.